### DISCUSSION TITLE
Csla.Blazor.ViewModel and firing specific rules on Save rather than OnModelChanging

### DISCUSSION THREAD

**Discussion by adrianwright109** [MarimerLLC/csla#2184] [Questions]
Hi,

I am using the `Csla.Blazor.ViewModel` (aka `vm`) for my `ClientEdit` object on a razor page.

I'm subscribing to the `ModelPropertyChanged` event on the `Csla.Blazor.ViewModel` to call `StateHasChanged()` this is allowing the `BusinessRule`s on the `ClientEdit` object to fire and update the changed `Property` validation warnings, which in turn only enables the `Save` button when all fields are valid based on the `vm.Model.IsSavable` `Property`.

This is all working great.

The issue I'm facing is that one of the `BusinessRule`s I've added to the `ClientName` `Property` isn't very _lightweight_ as its making a database call to determine if the `ClientName` value already exists in the database as they are required to be unique within the application.

This wouldn't be an issue except for the fact that this `BusinessRule` like the other `BusinessRule`s are being triggered on every character change (for a textbox).

Is it possible to exclude certain rules so that they only get checked when the `ClientEdit` object is saved not on the `ModelPropertyChanged` event?

**Comment by rockfordlhotka** [comment]
@jonnybee can probably offer the most in-depth answer.

I assume you know about rule priorities, so this long-running rule only runs if the cheaper rules all succeed?

Normally what I'd do, like in WPF or something, is have this particular UI control only update the business object property when the user tabs out, not on keypress. You can do that with Blazor also. This _usually_ solves the issue and is probably the easiest answer.

I know that @jonnybee added some more advanced capabilities to the rules engine, where the context includes info about _why_ the rule is being executed. I don't recall if that allows you do tell the difference between the rule executing due to a property change vs a `CheckRules` call.

And that's what you are really talking about here. You'd need to override `SaveAsync` and call `CheckRules` to trigger the rules to run on save, and have your long-running rule be aware that it should _not_ run normally, but should only run when `CheckRules` is invoked.

**Comment by richardbartley** [comment]
There were some RunMode options to control when the rule might fire.

```
BusinessRules.AddRule(new NotOptionDependencyRulesBroken { PrimaryProperty = SelectedOptionsProperty,
                                                        AffectedProperties = { BrokenOptionDependencyRulesProperty, RevisedSelectedOptionsProperty }, 
                                                                        RunMode = Csla.Rules.RunModes.DenyAsAffectedProperty,
                                                                        Priority = 10 });
```

Alternatively, in the past I have suspended the rule by using another property to control the rule logic, and then enable it again in the Save method by un-suspending the rule and calling CheckRules manually if needed.
